typing/includemod.ml

```diff
@@ -427,29 +426,49 @@ type core_relation = {
 }
 
 (**
-   In the group of mutual functions below, the [~in_eq] argument is [true] when
-   we are in fact checking equality of module types.
-
-   The module subtyping relation [A <: B] checks that [A.T = B.T] when [A]
-   and [B] define a module type [T]. The relation [A.T = B.T] is equivalent
-   to [(A.T <: B.T) and (B.T <: A.T)], but checking both recursively would lead
-   to an exponential slowdown (see #10598 and #10616).
-   To avoid this issue, when [~in_eq] is [true], we compute a coarser relation
-   [A << B] which is the same as [A <: B] except that module types [T] are
-   checked only for [A.T << B.T] and not the reverse.
-   Thus, we can implement a cheap module type equality check [A.T = B.T] by
-   computing [(A.T << B.T) and (B.T << A.T)], avoiding the exponential slowdown
-   described above.
+   In the group of mutual functions below, the [~direction] argument tracks the
+   subtyping direction at the syntactic level.
+
+   The [posivity] is used in the [cmt_declaration_dependencies] to distinguish
+   between directed and undirected edges.
+
+   The [in_eq] field is [true] when we are checking both directions inside of
+   module types which allows optimizing module type equality checks. The module
+   subtyping relation [A <: B] checks that [A.T = B.T] when [A] and [B] define a
+   module type [T]. The relation [A.T = B.T] is equivalent to [(A.T <: B.T) and
+   (B.T <: A.T)], but checking both recursively would lead to an exponential
+   slowdown (see #10598 and #10616). To avoid this issue, when [in_eq] is
+   [true], we compute a coarser relation [A << B] which is the same as [A <: B]
+   except that module types [T] are checked only for [A.T << B.T] and not the
+   reverse. Thus, we can implement a cheap module type equality check [A.T =
+   B.T] by computing [(A.T << B.T) and (B.T << A.T)], avoiding the exponential
+   slowdown described above.
 *)
+type positivity = Negative | Positive | Strictly_positive
+type directionality = { in_eq : bool; positivity : positivity }
```

Comment:
@gasche , would you mind having a look at this part of the PR to check if it is understandable? (since the authorship of this part of the PR has become fuzzy between me and @voodoos )

Reply:
I wonder what is the current exchange rate between Merlin PRs and Pattern-Matching Bug PRs.

I'll have a look.